File: src/Main/java/com/nishant/algorithms/Sorting/HeapSort/HeapSort.java
package com.nishant.algorithms.Sorting.HeapSort;

import java.util.Arrays;

public class HeapSort {

  private static int heapSize;

  public static void sort(int[] arr) {
    buildHeap(arr);
    while (heapSize > 0) {
      arr[heapSize - 1] = popMax(arr);
    }
  }

  private static void buildHeap(int[] arr) {
    heapSize = arr.length;

    for (int i = (arr.length >> 1) - 1; i >= 0; i--) {
      heapify(arr, i);
    }
  }

  private static void heapify(int[] arr, int i) {
    int left = getLeft(i), right = getRight(i), largest = i;

    if (left < heapSize && arr[left] > arr[largest]) {
      largest = left;
    }

    if (right < heapSize && arr[right] > arr[largest]) {
      largest = right;
    }

    if (i != largest) {
      swap(arr, i, largest);
      heapify(arr, largest);
    }
  }

  private static int popMax(int[] heap) {
    int toReturn = heap[0];
    swap(heap, 0, heapSize - 1);
    heapSize--;
    heapify(heap, 0);
    return toReturn;
  }

  private static int getLeft(int i) {
    return (i << 1) + 1;
  }

  private static int getRight(int i) {
    return (i << 1) + 2;
  }

  private static void swap(int[] arr, int pos1, int pos2) {
    int temp = arr[pos1];
    arr[pos1] = arr[pos2];
    arr[pos2] = temp;
  }

  public static void main(String[] args) {
    int[] arr = {1, 5, 3, 5, 3, 2, 4, 5, 2};
    System.out.println(Arrays.toString(arr));
    sort(arr);
    System.out.println(Arrays.toString(arr));
  }
}

File: src/main/python/sorting/heapsort/heap_sort.py
heap_size = 0


def heap_sort(to_sort):
    build_heap(to_sort)
    while heap_size > 0:
        pop_max(to_sort)


def build_heap(to_sort):
    global heap_size
    heap_size = len(to_sort)

    for i in range((heap_size >> 1) - 1, -1, -1):
        heapify(to_sort, i)


def pop_max(to_sort):
    global heap_size
    heap_size -= 1
    to_sort[0], to_sort[heap_size] = to_sort[heap_size], to_sort[0]
    heapify(to_sort, 0)


def heapify(to_sort, i):
    left, right, max_num = get_left(i), get_right(i), i

    if left < heap_size and to_sort[left] > to_sort[max_num]:
        max_num = left

    if right < heap_size and to_sort[right] > to_sort[max_num]:
        max_num = right

    if max_num != i:
        to_sort[i], to_sort[max_num] = to_sort[max_num], to_sort[i]
        heapify(to_sort, max_num)


def get_left(i):
    return (i << 1) + 1


def get_right(i):
    return (i << 1) + 2


def get_parent(i):
    return (i >> 1) if i % 2 == 1 else (i >> 1) - 2


if __name__ == "__main__":
    arr = [1, 2, 3, 6, 4, 5]
    print(str(arr))
    heap_sort(arr)
    print(str(arr))
